security/pkg/nodeagent/sds/sdsservice.go

```diff
@@ -421,24 +444,49 @@ func pushSDS(con *sdsConnection) error {
 	if err = con.stream.Send(response); err != nil {
 		sdsServiceLog.Errorf("Failed to send response for SDS resource %q to proxy connection %q: %v",
 			resourceName, conID, err)
+		sdsMetrics.pushFailurePerConn.With(prometheus.Labels{"resourcePerConn": resourceName + "-" + conID}).Inc()
 		return err
 	}
 
+	// Update metrics after push to avoid adding latency to SDS push.
 	if secret.RootCert != nil {
 		sdsServiceLog.Infof("Pushed root cert (resource name: %q) from node agent to proxy connection: %q\n",
 			resourceName, conID)
 		sdsServiceLog.Debugf("Pushed root cert %+v (resource name: %q) to proxy connection: %q\n",
 			string(secret.RootCert), resourceName, conID)
+		sdsMetrics.rootCertExpiryTimestamp.With(prometheus.Labels{"resourcePerConn": resourceName + "-" + conID}).Set(
+			extractCertExpiryTimestamp(resourceName, conID, secret.RootCert))
 	} else {
 		sdsServiceLog.Infof("Pushed key/cert pair (resource name: %q) from node agent to proxy: %q\n",
 			resourceName, conID)
 		sdsServiceLog.Debugf("Pushed certificate chain %+v (resource name: %q) to proxy connection: %q\n",
 			string(secret.CertificateChain), resourceName, conID)
+		sdsMetrics.serverCertExpiryTimestamp.With(prometheus.Labels{"resourcePerConn": resourceName + "-" + conID}).Set(
+			extractCertExpiryTimestamp(resourceName, conID, secret.CertificateChain))
 	}
-
+	sdsMetrics.pushPerConn.With(prometheus.Labels{"resourcePerConn": resourceName + "-" + conID}).Inc()
+	sdsMetrics.pendingPushPerConn.With(prometheus.Labels{"resourcePerConn": resourceName + "-" + conID}).Desc()
+	sdsMetrics.totalPush.Inc()
 	return nil
 }
 
+// extractCertExpiryTimestamp returns the cert expiration time as unix timestamp.
+func extractCertExpiryTimestamp(resourceName, conID string, certByte []byte) float64 {
```

Comment:
could we just use ```secret.ExpireTime``` instead of introducing this func ?

Reply:
@quanjielin Thanks for the advice. The secret.ExpireTime is the leaf cert expire time. We don't extract root cert expire time and pass that into ```model.SecretItem``` for root cert. Shall we do this
(1) Examine all expire time in the cert chain. https://github.com/istio/istio/blob/482f908e2cde78adb73ec599d16cf6186f07ce2b/security/pkg/nodeagent/cache/secretcache.go#L631
(2) Extract expire time and pass into ```model.SecretItem``` for root cert.